### DISCUSSION TITLE
Is it possible to create a LUA subclass form an exported C++ class?

### DISCUSSION THREAD

**Discussion by NI-LAm** [gengyong/luaaa#24] [Q&A] [answered]
using `example.lua` and [LUA tutorial](https://www.lua.org/pil/16.2.html) I tried to subclass AwesomeCat like so

    SpecialCat = AwesomeCat:new("special")
    function SpecialCat:onlyInSpecial()
        print(self.getName() .. " has a special cat function")
    end
    function SpecialCat:speak(text)
        print("Special cat says: " .. text)
    end

This should add a new function `onlyInSpecial` to SpecialCat and overwrite `speak` but that does not seem to work. LUA refuses `AwesomeCat:new` with

    lua err: [string "console"]:60: calling 'new' on bad self (string expected, got table)

using `AwesomeCat.new` works, but then I am not allowed to add new functions to SpecialCat

    lua err: [string "console"]:61: attempt to index a AwesomeCat value (global 'SpecialCat')

Is this possible at all and if so how to do this?

**Comment by gengyong** [comment, marked as answer]
in luaaa, exported class is userdata instead of table:

```lua
print("type of AwesomeCat:", type(AwesomeCat.new("CCC")))
```
output:
```
type of AwesomeCat:   userdata
```

we can wrapper `userdata` to `table` to make inheritance works, likes below:

```lua
SpecialCat = {}

function SpecialCat:new(name)
	o = AwesomeCat:new(name)
	setmetatable(self, getmetatable(o))
	return self
end
```

but there is an issue in luaaa codes, luaaa always expect an userdata from stack, so we had to store origin userdata in table, and read it from luaaa by specified name, we use '@' as it's name:

```lua
function SpecialCat:new(name)
	o = AwesomeCat:new(name)
	setmetatable(self, getmetatable(o))
	self['@'] = o
	return self
end
```

now wen can extend SpecialCat and override it's base class's method:
```lua
function SpecialCat:onlyInSpecial()
    print(self:getName() .. " has a special cat function")
end

function SpecialCat:speak(text)
    print("Special cat " .. self:getName() .. " says: " .. text)
    -- and base method can be call as below:
    self["@"]:speak(text)
end
``` 

rearrange above codes, we got below:
```lua
-- put utility funtions to name scope 'luaaa'
luaaa = {}

-- create subclass for base, obj can be exist table or nil
function luaaa:extend(base, obj)
	derived = obj or {}
	derived.new = function(self, ...)
		o = base.new(...)
		setmetatable(self, getmetatable(o))
		self["@"] = o
		return self
	end
	return derived
end

-- get base class of obj
function luaaa:base(obj)
	if (type(obj) == "table") then
		return obj["@"]
	end
	return nil
end

```

by with above utility functions, you can extend lua class likes below:

```lua
SpecialCat = luaaa:extend(AwesomeCat, {value = 1})
-- or:
--   SpecialCat = luaaa:extend(AwesomeCat)
-- in this case there no attribute for extended new class

function SpecialCat:onlyInSpecial()
    print(self:getName() .. " has a special cat function")
    print("Special cat " .. self:getName() .." has value:" .. self.value)
end

function SpecialCat:speak(text)
    print("Special cat[" .. self:getName() .. "] says: " .. text)
    -- call override base method:
	luaaa:base(self):speak(text)
end
```

use SpecialCat:
```lua
xxx = SpecialCat:new("xxx")
xxx:speak("I am a special cat.")
xxx:onlyInSpecial()
```


BTW: 
as you mentioned, the 'new()' operator of exported lua class should be called with "." instead of ":".
as you know, "obj:new(...)" is equal to "obj.new(self, ...)", 'self' is 'obj' itself. 
in luaaa, 'self' was omitted in constructor method because it's not required in luaaa before, and I misunderstand the lifecycle of lua object, it seems 'obj:new()' is more make sense. this behavior is easy to change, but this change will break exists lua codes, so I will change it in next major version.

**Comment by gengyong** [comment]
please pull the latest version(>= 0.2.3), then add below utility functions to your project .
```lua
-- put utility funtions to name scope 'luaaa'
luaaa = {}

-- create subclass for base, obj can be exist table or nil
function luaaa:extend(base, obj)
	derived = obj or {}
	derived.new = function(self, ...)
                -- NOTICE: in next major version, below line will be changed to 'o = base:new(...)'
		o = base.new(...)
		setmetatable(self, getmetatable(o))
		self["@"] = o
		return self
	end
	return derived
end

-- get base class of obj
function luaaa:base(obj)
	if (type(obj) == "table") then
		return obj["@"]
	end
	return nil
end
```
